### PR TITLE
feat(connector): add tdf renderer connector

### PR DESCRIPTION
**Pull request by canluhuang**
# Pre-PR Checklist

- [x] I added/updated relevant documentation.
- [x] I followed the [Convention Commit](https://conventionalcommits.org/) guideline with maximum 72 characters to submit commit message.
- [x] I squashed the repeated code commits.
- [x] I signed the [CLA].
- [x] I added/updated test cases to check the change I am making.
- [x] All existing and new tests are passing.

### CODE REVIEW

File: framework/android/src/main/cpp/CMakeLists.txt [hunk 1]
```diff
@@ -81,6 +81,14 @@ add_subdirectory(${PROJECT_ROOT_DIR}/framework/android/connector/renderer/native
 target_link_libraries(${PROJECT_NAME} PUBLIC native_renderer_connector)
 # endregion
 
+# region tdf renderer delegate
+if (${ENABLE_TDF_RENDER} STREQUAL "true")
```
Comment (churchill-zhang) [inline framework/android/src/main/cpp/CMakeLists.txt:85]: 加了 tdf render，上面的 native render 也要有个开关了

Comment (canluhuang) [reply framework/android/src/main/cpp/CMakeLists.txt:85]: 因为目前 tdf render 仍会依赖到 native-renderer，主要有两点：
1、TDFRender 继承 RenderProxy ，RenderProxy 目前是定义在 native-renderer 模块
2、现存的自定义 View 需要依赖  native-render 提供渲染

所以目前 natiev-renderer 还是必须引入的，开关是不是可以先不加？

针对第1点，RenderProxy 等相关类本来是抽到 renderer-support 模块，这样可以降低下耦合，但发现 FrameworkProxy 依赖 FontAdapter、ImageLoaderAdapter 也都是 native-render 的，改动太大，不好 Review，就准备后面再处理

**Review by zoomchan-cxj — APPROVED** [review]

**Review by medns — CHANGES_REQUESTED** [review]
rebase upstream & use new dependency describe method

File: framework/android/build.gradle [hunk 2]
```diff
@@ -73,7 +73,8 @@ android {
                             "-DANDROID_STL=$ANDROID_STL",
                             "-DV8_COMPONENT=$V8_COMPONENT",
                             "-DTDF_CORE_VERSION=$TDF_CORE_VERSION",
-                            "-DHIDDEN_LIBRARY_SYMBOL=$HIDDEN_LIBRARY_SYMBOL"
+                            "-DHIDDEN_LIBRARY_SYMBOL=$HIDDEN_LIBRARY_SYMBOL",
+                            "-DENABLE_TDF_RENDER=${ENABLE_TDF_RENDER}"
```
Comment (medns) [inline framework/android/build.gradle:77]: theoretically, no longer needed `ENABLE_TDF_RENDER` flag

**Review by medns — CHANGES_REQUESTED** [review]

File: framework/android/build.gradle (hunk 2 above)
Comment (vimerzhao) [reply framework/android/build.gradle]: fixed

**Review by ilikethese — APPROVED** [review]

**Comment by medns** [comment]
@siguangli 👀

File: framework/examples/android-demo/build.gradle [hunk 3]
```diff
@@ -73,6 +73,11 @@ dependencies {
         compileOnly project(path: ':hippy-support')
         compileOnly project(path: ':renderer-native')
     }
+
+    if (MODULES.contains("renderer/tdf")) {
+        api "androidx.exifinterface:exifinterface:1.3.3"
+        api "com.tencent.tdfcore:embedder:0.0.4"
+    }
```
[the remark this answers is not in the document]
Comment (vimerzhao) [reply framework/examples/android-demo/build.gradle]: fixed

File: settings.gradle [hunk 4]
```diff
@@ -9,6 +9,11 @@ project(':connector-dom').projectDir = new File('framework/android/connector/dom
 include 'renderer-native'
 project(':renderer-native').projectDir = new File('renderer/native/android')
 
+if (MODULES.contains("renderer/tdf")) {
+    include 'renderer-tdf'
```
[the remark this answers is not in the document]
Comment (vimerzhao) [reply settings.gradle]: fixed

**Review by vimerzhao — DISMISSED** [review]

**Review by medns — CHANGES_REQUESTED** [review]
rebase upstream & use new dependency describe method

File: framework/android/connector/renderer/tdf/gradle.properties [hunk 5]
```diff
@@ -0,0 +1,8 @@
+android.enableJetifier=true
+android.useAndroidX=true
```
Comment (medns) [inline framework/android/connector/renderer/tdf/gradle.properties:2]: unnecessary define

File: framework/android/connector/renderer/tdf/gradle.properties [hunk 6]
```diff
@@ -0,0 +1,8 @@
+android.enableJetifier=true
```
Comment (medns) [inline framework/android/connector/renderer/tdf/gradle.properties:1]: missing license header

File: framework/android/connector/support/build.gradle [hunk 7]
```diff
@@ -38,5 +38,8 @@ android {
 }
 
 dependencies {
+    compileOnly project(path: ':renderer-native')
```
Comment (medns) [inline framework/android/connector/support/build.gradle:41]: @siguangli

File: renderer/support/build.gradle [hunk 8]
```diff
@@ -0,0 +1,45 @@
+apply plugin: 'com.android.library'
```
Comment (medns) [inline renderer/support/build.gradle:1]: why need renderer/support project?

**Review by medns — CHANGES_REQUESTED** [review]

File: renderer/support/build.gradle (hunk 8 above)
Comment (canluhuang) [reply renderer/support/build.gradle]: This is unnecessary and has been removed

File: framework/android/connector/support/src/main/java/com/openhippy/connector/RenderConnector.java [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+import android.view.ViewGroup;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import com.tencent.mtt.hippy.common.Callback;
+import com.tencent.renderer.FrameworkProxy;
+import com.tencent.renderer.RenderProxy;
+import java.util.List;
+import java.util.Map;
+
+public abstract class RenderConnector implements Connector, RenderProxy {
```
Comment (siguangli) [inline framework/android/connector/support/src/main/java/com/openhippy/connector/RenderConnector.java:30]: connector support下面不应该对 renderer工程有依赖，这里不能去实现renderProxy

File: framework/android/connector/support/build.gradle (hunk 7 above)
Comment (siguangli) [inline framework/android/connector/support/build.gradle:41]: connector support下面不应该对 renderer工程有依赖

File: framework/android/connector/support/build.gradle [hunk 10]
```diff
@@ -38,5 +38,8 @@ android {
...
+    compileOnly project(path: ':hippy-support')
```
Comment (siguangli) [inline framework/android/connector/support/build.gradle:42]: connector support下面不应该对hippy-support 依赖

File: framework/android/connector/support/src/main/java/com/openhippy/connector/RenderConnector.java (hunk 9 above)
Comment (canluhuang) [reply framework/android/connector/support/src/main/java/com/openhippy/connector/RenderConnector.java]: fixed

File: framework/android/connector/support/build.gradle (hunk 7 above)
Comment (canluhuang) [reply framework/android/connector/support/build.gradle]: fixed

File: framework/android/connector/support/build.gradle (hunk 10 above)
Comment (canluhuang) [reply framework/android/connector/support/build.gradle]: fixed

File: framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java [hunk 11]
```diff
@@ -42,84 +43,81 @@ public NativeRenderer() {
...
             return mRenderer.replaySnapshot(context, snapshotMap);
         }
         return null;
     }
 
-    public void setFrameworkProxy(@NonNull FrameworkProxy proxy) {
+    @Override public void setFrameworkProxy(@NonNull Object proxy) {
         if (mRenderer != null) {
-            mRenderer.setFrameworkProxy(proxy);
+            mRenderer.setFrameworkProxy((FrameworkProxy) proxy);
```
Comment (siguangli) [inline framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java:81]: 这里传进来的Object对象最好还是判断下  instance of FrameworkProxy

File: framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java [hunk 12]
```diff
@@ -42,84 +43,81 @@ public NativeRenderer() {
...
-            mRenderer.onRuntimeInitialized(root.getId());
+            mRenderer.onRuntimeInitialized(rootId);
         }
     }
 
-    public void recordSnapshot(@NonNull View rootView, @NonNull final Callback<byte[]> callback) {
+    @Override public void recordSnapshot(int rootId, @NonNull Object callback) {
         if (mRenderer != null) {
-            mRenderer.recordSnapshot(rootView.getId(), callback);
+            mRenderer.recordSnapshot(rootId, (Callback<byte[]>)callback);
```
Comment (siguangli) [inline framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java:60]: 这里传进来的Object对象最好还是判断下 instance of Callback<byte[]

File: framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java [hunk 13]
```diff
@@ -42,84 +43,81 @@ public NativeRenderer() {
         }
     }
 
-    public void destroyRoot(int rootId) {
+    @Override public void destroyRoot(int rootId) {
```
Comment (siguangli) [inline framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java:46]: @Override 为什么会和public写在一行，代码格式不对吧

File: framework/android/connector/renderer/tdf/gradle.properties (hunk 5 above)
Comment (canluhuang) [reply framework/android/connector/renderer/tdf/gradle.properties]: fixed

File: framework/android/connector/renderer/tdf/gradle.properties (hunk 6 above)
Comment (canluhuang) [reply framework/android/connector/renderer/tdf/gradle.properties]: fixed

File: framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java (hunk 11 above)
Comment (canluhuang) [reply framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java:87]: fixed

File: framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java (hunk 12 above)
Comment (canluhuang) [reply framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java]: fixed

File: framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java (hunk 13 above)
Comment (canluhuang) [reply framework/android/connector/renderer/native/src/main/java/com/openhippy/connector/NativeRenderer.java]: fixed

**Review by siguangli — APPROVED** [review]

File: framework/android/connector/renderer/tdf/build.gradle [hunk 14]
```diff
@@ -0,0 +1,49 @@
+apply plugin: 'com.android.library'
```
Comment (medns) [inline framework/android/connector/renderer/tdf/build.gradle:1]: missing licenses headers

File: renderer/tdf/android/src/main/java/com/tencent/renderer/TDFRenderer.java [hunk 15]
```diff
@@ -114,12 +120,38 @@ public void init(@Nullable List<Class<?>> controllers, @Nullable ViewGroup rootV
         }
     }
 
+    @Override public void destroyRoot(int rootId) {
```
Comment (medns) [inline renderer/tdf/android/src/main/java/com/tencent/renderer/TDFRenderer.java:123]: `@Override` annotation move to previous line

File: renderer/tdf/common/include/renderer/tdf/tdf_render_manager.h [hunk 16]
```diff
@@ -25,6 +25,7 @@
 #include "dom/dom_node.h"
 #include "dom/render_manager.h"
 #include "footstone/serializer.h"
+#include "vfs/uri_loader.h"
```
Comment (medns) [inline renderer/tdf/common/include/renderer/tdf/tdf_render_manager.h:28]: missing blank line, between `#include` and `#pragma`

**Review by medns — CHANGES_REQUESTED** [review]

File: renderer/tdf/common/include/renderer/tdf/tdf_render_manager.h (hunk 16 above)
Comment (canluhuang) [reply renderer/tdf/common/include/renderer/tdf/tdf_render_manager.h:28]: fixed

File: renderer/tdf/android/src/main/java/com/tencent/renderer/TDFRenderer.java (hunk 15 above)
Comment (canluhuang) [reply renderer/tdf/android/src/main/java/com/tencent/renderer/TDFRenderer.java]: fixed

File: framework/android/connector/renderer/tdf/build.gradle (hunk 14 above)
Comment (canluhuang) [reply framework/android/connector/renderer/tdf/build.gradle:23]: fixed

**Review by medns — APPROVED** [review]
LGTM

**Review by siguangli — APPROVED** [review]

**Review by vimerzhao — DISMISSED** [review]

File: framework/android/connector/renderer/tdf/src/main/cpp/src/tdf_renderer_jni.cc [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+  auto dom_manager_id = footstone::check::checked_numeric_cast<jint, uint32_t>(j_dom_manager_id);
+  std::any dom_manager;
+  flag = hippy::global_data_holder.Find(dom_manager_id, dom_manager);
+  FOOTSTONE_CHECK(flag);
+  auto dom_manager_object = std::any_cast<std::shared_ptr<DomManager>>(dom_manager);
+  render_manager_object->SetDomManager(dom_manager_object);
+}
+
+}
+}
```
Comment (ilikethese) [inline framework/android/connector/renderer/tdf/src/main/cpp/src/tdf_renderer_jni.cc:88]: namespace 没有注释

File: renderer/tdf/android/src/main/cpp/include/render/tdf_render_bridge.h [hunk 18]
```diff
@@ -22,12 +22,21 @@
...
-  static void RegisterScopeForUriLoader(uint32_t render_id, const std::shared_ptr<hippy::driver::Scope>& scope);
-};
+void RegisterTDFEngine(JNIEnv *j_env, jobject j_obj, jint j_render_id,
+                       jlong j_engine_id, jint j_root_view_id);
+
+void SetUriLoader(JNIEnv *j_env, jobject j_obj, jint j_render_id, jint j_vfs_id);
+
+}
+}
+}
```
Comment (ilikethese) [inline renderer/tdf/android/src/main/cpp/include/render/tdf_render_bridge.h:42]: namespace 注释

File: renderer/tdf/common/src/renderer/tdf/tdf_render_manager.cc [hunk 19]
```diff
@@ -99,8 +101,29 @@ TDFRenderManager::TDFRenderManager() : RenderManager("TDFRenderManager") {
 }
 
 void TDFRenderManager::RegisterShell(uint32_t root_id, const std::shared_ptr<tdfcore::Shell>& shell) {
-  auto root_node = TDF_MAKE_SHARED(RootViewNode, hippy::DomNode::RenderInfo{root_id, 0, 0}, shell, GetDomManager(),
-                                   GetUriDataGetter());
+  auto uri_data_getter =  [this](const string_view &uri, const RootViewNode::DataCb& callback) {
```
Comment (ilikethese) [inline renderer/tdf/common/src/renderer/tdf/tdf_render_manager.cc:104]: 是不是weak_this 更好

**Review by ilikethese — CHANGES_REQUESTED** [review]
LGTM

File: framework/android/connector/renderer/tdf/src/main/cpp/src/tdf_renderer_jni.cc (hunk 17 above)
Comment (canluhuang) [reply framework/android/connector/renderer/tdf/src/main/cpp/src/tdf_renderer_jni.cc]: fixed

File: renderer/tdf/android/src/main/cpp/include/render/tdf_render_bridge.h (hunk 18 above)
Comment (canluhuang) [reply renderer/tdf/android/src/main/cpp/include/render/tdf_render_bridge.h:42]: fixed

File: renderer/tdf/common/src/renderer/tdf/tdf_render_manager.cc (hunk 19 above)
Comment (canluhuang) [reply renderer/tdf/common/src/renderer/tdf/tdf_render_manager.cc]: 确实，已优化

File: renderer/tdf/android/src/main/cpp/include/render/tdf_render_bridge.h (hunk 18 above)
Comment (ilikethese) [reply renderer/tdf/android/src/main/cpp/include/render/tdf_render_bridge.h:42]: 好像漏了这个

**Review by ilikethese — CHANGES_REQUESTED** [review]

File: renderer/tdf/android/src/main/cpp/include/render/tdf_render_bridge.h (hunk 18 above)
Comment (canluhuang) [reply renderer/tdf/android/src/main/cpp/include/render/tdf_render_bridge.h]: 已补齐

**Review by ilikethese — APPROVED** [review]
LGTM

**Review by siguangli — APPROVED** [review]

**Review by vimerzhao — APPROVED** [review]

**Review by churchill-zhang — APPROVED** [review]
LGTM

**Pull request merged**